### PR TITLE
improve async iter performance

### PR DESCRIPTION
**Pull request by zsoltm**
### Description

Performance fix for async iter. Use chunked fetch approach, chunk size can be altered from the outside.

Fixes:  #34

### CODE REVIEW

File: aiosqlite/cursor.py
Comment (jreese) [inline aiosqlite/cursor.py:15]: Please don't use double underscore prefixes. This causes name mangling that both makes it difficult to debug from tracebacks and also makes it more difficult when subclassing.

File: aiosqlite/cursor.py
Comment (jreese) [inline aiosqlite/cursor.py:16]: This seems like it's potentially a lot to buffer in memory for each cursor. Did you do performance tuning/testing to arrive at this value, or was it arbitrarily chosen? I think it would also be useful to have this as an optional property set on the Connection class, so that it could be set once and reused for each cursor on that connection. I would also probably default to a smaller value, like 64, since it's possible that each row could be quite large.

File: aiosqlite/cursor.py
Comment (jreese) [inline aiosqlite/cursor.py:38]: Overall, I think all of this could be more cleanly implemented using an async generator. This removes the need for `self.__chunk`, and also reduces the complexity around handling/raising stop iteration, etc. Something like:

```python
def __aiter__(self):
    return self._fetch_chunked()

async def _fetch_chunked(self):
    while True:
        rows = await self.fetchmany(...)
        for row in rows:
            yield row
```

File: aiosqlite/tests/smoke.py
Comment (jreese) [inline aiosqlite/tests/smoke.py:139]: This would be a better fit in `tests/perf.py`.

**Review by jreese — CHANGES_REQUESTED** [review]
I really appreciate you tackling this. This is a great start, but there are some changes I'd like to see before merging. Cheers!

**Comment by zsoltm** [comment]
@jreese thank you for your feedback and review, I'll apply your suggestions as soon as I find some time, hopefully within a few days.

File: aiosqlite/cursor.py
Comment (zsoltm) [reply aiosqlite/cursor.py]: Not really, it was just a guess, but I agree with your concerns, and in fact 64 turns out to be a nice balance between RAM / performance.

On my modest hardware I got these perf results (the original did 36k iters / 2s):

Chunk size | Iterations | Duration | Rate | Gain | Gain over the original
-- | -- | -- | -- | -- | --
16 | 410,705 | 2.0s | 205,317 |   | ×11
32 | 644,097 | 2.0s | 321,992 | 57% | ×18
64 | 893,560 | 2.0s | 446,702 | 39% | ×24
128 | 1,111,297 | 2.0s | 555,551 | 24% | ×30
256 | 1,275,905 | 2.0s | 637,841 | 15% | ×35
512 | 1,380,353 | 2.0s | 690,056 | 8% | ×38
1024 | 1,447,937 | 2.0s | 723,842 | 5% | ×39

File: aiosqlite/cursor.py
Comment (zsoltm) [reply aiosqlite/cursor.py]: Indeed way better this way, also performance wise. Sometimes it's hard to remember all this pythonic coolness coming from a different language.

File: aiosqlite/cursor.py
Comment (jreese) [inline aiosqlite/cursor.py:21]: I would still prefer to drop the second underscore, and just have it named `_fetch_chunked`.

**Review by jreese — CHANGES_REQUESTED** [review]
One last change I'd like to see, but everything else looks great! Happy to merge this once that last bit is fixed.

**Review by jreese — APPROVED** [review]

**Comment by jreese** [comment]
Thank you for working through this!

**Pull request merged**
